OpenRA.Mods.Common/Traits/ProductionIconOverlay.cs

```diff
@@ -0,0 +1,137 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using OpenRA.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Attach this to the player actor. When attached, enables all actors possessing the LevelupWhenCreated ",
+		"trait to have their production queue icons render with an overlay defined in this trait. ",
+		"The icon change occurs when LevelupWhenCreated.Prerequisites are met.")]
+	public class ProductionIconOverlayInfo : ITraitInfo, Requires<TechTreeInfo>
+	{
+		[Desc("Image used for the overlay.")]
+		public readonly string Image = "rank";
+
+		[Desc("Sequence used for the overlay (cannot be animated).")]
+		public readonly string Sequence = "rank";
+
+		[Desc("Palette to render the sprite in. Reference the world actor's PaletteFrom* traits.")]
+		public readonly string Palette = "effect";
+
+		[Desc("Point on the production icon's used as reference for offsetting the overlay. ",
+			"Possible values are any combination of Top, VCenter, Bottom and Left, HCenter, Right separated by a comma.")]
+		public readonly ReferencePoints ReferencePoint = ReferencePoints.Top | ReferencePoints.Right;
+
+		[Desc("Pixel offset relative to the icon's reference point.")]
+		public readonly int2 Offset = int2.Zero;
+
+		[Desc("Visual scale of the overlay.")]
+		public readonly float Scale = 2f;
+
+		public object Create(ActorInitializer init) { return new ProductionIconOverlay(init, this); }
+	}
+
+	public class ProductionIconOverlay : ITechTreeElement
+	{
+		public readonly Sprite Sprite;
+		public readonly string Palette;
+		public readonly float Scale;
+
+		const string Prefix = "ProductionIconOverlay.";
```

Comment:
In that case I'd say this calls for a 
`// HACK: ...` with an explanation.
Because looking at the code in this trait, this is pretty redundant for the keys.